### PR TITLE
RDOIAT-522 - add changeFocusedOption method

### PR DESCRIPTION
**Pull request by OS-ruineves**
Added changeFocusedOption that allows the dropdown on SS to focus perfect matchs when searching

### CODE REVIEW

File: src/Select.js [hunk 1]
```diff
@@ -562,6 +562,32 @@ export default class Select extends Component<Props, State> {
     });
   }
 
+  changeFocusedOption(opt: string){
```
Comment (os-lmo) [inline src/Select.js:565]: Why not just `focusOption` ?

File: src/Select.js [hunk 2]
```diff
@@ -562,6 +562,32 @@ export default class Select extends Component<Props, State> {
...
+    if(opt == null){
+       return;
+    }
+    const { menuOptions } = this.state;
+    const options = menuOptions.focusable;
+    
+    if(!options.length){
+       return;
+    }
+    let index = -1;
```
Comment (os-lmo) [inline src/Select.js:575]: Add empty line after the `if`

File: src/Select.js [hunk 3]
```diff
@@ -562,6 +562,32 @@ export default class Select extends Component<Props, State> {
...
+    let i = 0;
+    for(i = 0; i < options.length; i++){
+      if(options[i].label == opt){
+        index = i;
+        break;
+      }
+    }
+    if(index == -1){
```
Comment (os-lmo) [inline src/Select.js:583]: Same here

File: src/Select.js [hunk 4]
```diff
@@ -562,6 +562,32 @@ export default class Select extends Component<Props, State> {
...
+      return;
+    }
+    this.setState({
```
Comment (os-lmo) [inline src/Select.js:586]: And here

File: src/Select.js [hunk 5]
```diff
@@ -562,6 +562,32 @@ export default class Select extends Component<Props, State> {
...
+    }
+    const { menuOptions } = this.state;
+    const options = menuOptions.focusable;
+    
+    if(!options.length){
+       return;
+    }
+    let index = -1;
+    let i = 0;
+    for(i = 0; i < options.length; i++){
```
Comment (os-lmo) [inline src/Select.js:577]: Suggestion: findIndex already does this - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex - providing a predicate as argument

Comment (os-lmo) [reply src/Select.js:577]: On the other side, why aren't we using the `find` function? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find .

**Review by os-lmo — COMMENTED** [review]
See my comments

File: src/Select.js (hunk 2 above)
Comment (osplm) [reply src/Select.js:575]: and spaces before the brackets

File: src/Select.js (hunk 1 above)
Comment (OS-ruineves) [reply src/Select.js:565]: There is a method already with that name that changes the focused option based on direction (up, down, last, etc)

File: src/Select.js (hunk 5 above)
Comment (OS-ruineves) [reply src/Select.js:577]: Right now I only send the label of the option to the method so `find` would not work, but `findIndex` is indeed the right way

Comment (os-lmo) [reply src/Select.js]: If you don't need the index but just the element, I would suggest using the `find`

File: src/Select.js [hunk 6]
```diff
@@ -562,6 +562,28 @@ export default class Select extends Component<Props, State> {
...
+    }
+
+    let optionToFocus = options.find(e => e.label == opt);
+    if(!optionToFocus) {
+      return;
+    }
+
+    this.setState({
+      focusedOption: optionToFocus,
+    })
```
Comment (os-lmo) [inline src/Select.js:584]: In the other `focusOption` function, we are also:
1. Changing the the state of `focusValue` to null
2. Setting ` this.scrollToFocusedOptionOnUpdate = true;`
https://github.com/OutSystems/react-select/blob/main/src/Select.js#L594

Aren't they needed here also?

**Comment by os-lmo** [comment]
Can you check if our pipeline is executing the automatic tests? In that case, can you create a test here - https://github.com/OutSystems/react-select/blob/main/src/__tests__/Select.test.js

File: src/Select.js (hunk 6 above)
Comment (OS-ruineves) [reply src/Select.js:586]: 1. focusedValue is used on multi select dropdowns, I had removed it while debugging, value operations change focusedOption to null, and option operations change focusedValue to null, so I will be adding it back again
2. yes, this should improve the experience on dropdowns with a bigger number of options

**Review by os-lmo — APPROVED** [review]

File: src/Select.js [hunk 7]
```diff
@@ -562,6 +562,30 @@ export default class Select extends Component<Props, State> {
...
+    }
+
+    const { menuOptions } = this.state;
+    const options = menuOptions.focusable;
+    
+    if(!options.length) {
+       return;
+    }
+
+    let optionToFocus = options.find(e => e.label == opt);
```
Comment (osplm) [inline src/Select.js:577]: this can be a const as well

File: src/Select.js (hunk 6 above)
Comment (osplm) [reply src/Select.js:586]: Also, what does the announceAriaLiveContext that is being called in other operations? Seems to update some kind of instructions. What are we losing not calling that here?

**Review by osplm — APPROVED** [review]

File: src/Select.js (hunk 6 above)
Comment (OS-ruineves) [reply src/Select.js:586]: announceAriaLiveContext and announceAriaLiveSelection are disabled in the web-components code, with the allowScreenReaders prop

**Pull request merged**